### PR TITLE
initial Upload API.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmrepo/305/

File: csclient/params/params.go [hunk 1]
```diff
@@ -449,3 +449,24 @@ type SetAuthCookie struct {
 	// Macaroons holds a slice of macaroons.
 	Macaroons macaroon.Slice
 }
+
+// UploadIdResponse holds the upload id response from a POST request to /upload.
+type UploadIdResponse struct {
+	UploadId string
+}
+
+// Parts holds the list of part for an upload to be completed.
```
Comment (rogpeppe) [inline csclient/params/params.go]: // Parts holds a list of all the parts that are required by a multipart
// upload, as required by a PUT request to /upload/$upload-id.

?

File: csclient/params/params.go [hunk 2]
```diff
@@ -449,3 +449,24 @@ type SetAuthCookie struct {
...
+type Parts struct {
+	Parts []Part
+}
+
+// Part represents one part of a multipart blob.
+type Part struct {
+	Hash string
+}
+
+// FinishUploadResponse holds the response to a put /upload/upload-id/part-number.
```
Comment (rogpeppe) [inline csclient/params/params.go]: // FinishUploadResponse holds the response to a put /upload/upload-id/part-number request.
?

File: csclient/params/params.go [hunk 3]
```diff
@@ -449,3 +449,24 @@ type SetAuthCookie struct {
...
+type FinishUploadResponse struct {
+	// Hash is the hash of the file.
```
Comment (rogpeppe) [inline csclient/params/params.go]: s/hash of the file/hash of the complete blob/

?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/charmrepo/306/

**Comment by fabricematrat** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmrepo-merge

**Pull request merged**